python/modules/IcePy/Types.cpp

```diff
@@ -3809,11 +3805,28 @@ IcePy::ExceptionWriter::ExceptionWriter(const PyObjectHandle& ex, const Exceptio
     }
 }
 
-IcePy::ExceptionWriter::~ExceptionWriter()
+IcePy::ExceptionWriter::ExceptionWriter(const ExceptionWriter& other)
 {
-    AdoptThread adoptThread; // Ensure the current thread is able to call into Python.
+    // Exception writer copy constructor can be called from a C++ thread when IcePy throws a ExceptionWriter object
+    // from the dispatch pipeline.
+
+    {
+        // Ensure the current thread is able to call into Python.
+        AdoptThread adoptThread;
+        _ex = other._ex;
+    }
+    _info = other._info;
+    _objects = other._objects;
+}
 
-    _ex = 0;
+IcePy::ExceptionWriter::~ExceptionWriter() noexcept
+{
+    // Exception writer destructor can be called from a C++ thread when IcePy throws a ExceptionWriter object
```

Comment:
I think this one is indeed ok because the exception thrown by ServantLocatorWrapper::locate/finished gets destroyed outside AdoptThread.